### PR TITLE
Revert defGetString() changes, use temporary memory context instead

### PR DESCRIPTION
**Pull request by hlinnaka**
See commit messages. With the temporary memory context, the memory usage is more or less the same as before. Without that, just reverting the defGetString() changes caused a 10% regression on the test case provided in the old JIRA ticket, MPP-23076.

I spent a long time trying a more ambitious approach, using the temporary memory context for the whole transformCreateStmt() call, and only copying the finished Query to MessageContext at the end. But that turned out to be difficult; there are all kinds of hacks that made that not work:
* Some CreateStmt and Partition* struct fields are missing copyObject() support. That's apparently not an active bug right now, if those structs are never copied, but we should fix those anyway.
* atpxPartAddList() assumes that CreateStmt->attr_encodings field is filled in by parse_analyze() in the original CreateStmt object, not the one that transformCreateStmt() returns.
* Even after fixing those, there were mysterious crashes from _out* functions with "make installcheck-good". I stopped pursuing that approach at that point, so I don't know what's going on, but possibly still more missing copy/out/read function support.
* it didn't actually seem to save much more memory than this limited patch.

So, this PR is less ambitious than that, but seems straightforward.

### CODE REVIEW

**Comment by hlinnaka** [comment]
On 11/03/16 15:56, Daniel Gustafsson wrote:
> Is there a case where `column` or `c->column` could be NULL here? I
> can't see one reading over the code but I'm not really familiar with
> this path. Since `equal()` has an additional NULL check before the
> `strcmp()` it would've passed clean.

Indeed there are: when c->deflt == TRUE, c->column is NULL. Regression 
tests caught that; I had not run them to completion before submitting 
the PR.

Updated the PR with a fixed version. There were some other bugs too: 
transformAttributeEncoding() used to scribble on the input CreateStmt, 
which didn't work too well when some parts of the CreateStmt were 
allocated in the temporary memory context, and the CreateStmt was reused 
later. But now it seems to work. Overall, this turned out to be much 
more complicated than I thought when I started hacking this, but having 
done all this, I think this is worth committing.

- Heikki

File: src/backend/parser/parse_partition.c [hunk 1]
```diff
@@ -2816,16 +2802,17 @@ validate_partition_spec(ParseState *pstate, CreateStmtContext *cxt,
...
+					foreach(alc, allPartNames)
 					{
-						ereport(ERROR,
-								(errcode(ERRCODE_INVALID_TABLE_DEFINITION),
-								 errmsg("duplicate partition name "
-										"for partition%s%s",
-										namBuf,
-										at_depth),
-							   parser_errposition(pstate, pElem->location)));
+						if (strcmp((char *) lfirst(alc), pElem->partName) == 0)
```
Comment (ashwinstar) [inline src/backend/parser/parse_partition.c]: strncmp

**Comment by ashwinstar** [comment]
:+1: 

@asimrp you have been in this area many times, can take a look.

File: src/backend/parser/parse_partition.c (hunk 1 above)
Comment (hlinnaka) [inline src/backend/parser/parse_partition.c]: On 16/03/16 03:26, Ashwin Agrawal wrote:
> strncmp

No, don't think so. We want to compare the partition name, not only a 
part of it. Also, the code this replaces uses list_member(), which uses 
equal() for comparing the String elements, which uses strcmp().

- Heikki

**Comment by danielgustafsson** [comment]
With the above mentioned potentials NULL's fixed I think this looks good. :+1:

**Comment by foyzur** [comment]
@ashwinstar do you guys wanna take ownership of this PR? Also, this change will diverge 4.3 too much from master. Do you have plan to backport? In other words, does it have less memory leak than the original fix? If I remember correctly, this one is related to the major leak fix because of rescanning.

**Comment by hlinnaka** [comment]
On 21/03/16 18:20, foyzur wrote:
> @ashwinstar do you guys wanna take ownership of this PR? Also, this
> change will diverge 4.3 too much from master. Do you have plan to
> backport? In other words, does it have less memory leak than the
> original fix? If I remember correctly, this one is related to the
> major leak fix because of rescanning.

The memory consumption is about the same with this patch. Probably 
depends somewhat on the statement. This is not a bugfix, just cleanup 
for the future, and should definitely not be backported.

- Heikki

**Comment by hlinnaka** [comment]
Pushed, thanks

**Pull request closed without merging**